common/crypto/sig_private_key.cpp

```diff
@@ -38,105 +39,157 @@
 /***END Conditional compile untrusted/trusted***/
 
 namespace pcrypto = pdo::crypto;
-namespace constants = pdo::crypto::constants;
-
-// Typedefs for memory management
-// Specify type and destroy function type for unique_ptrs
-typedef std::unique_ptr<BIO, void (*)(BIO*)> BIO_ptr;
-typedef std::unique_ptr<EVP_CIPHER_CTX, void (*)(EVP_CIPHER_CTX*)> CTX_ptr;
-typedef std::unique_ptr<BN_CTX, void (*)(BN_CTX*)> BN_CTX_ptr;
-typedef std::unique_ptr<BIGNUM, void (*)(BIGNUM*)> BIGNUM_ptr;
-typedef std::unique_ptr<EC_GROUP, void (*)(EC_GROUP*)> EC_GROUP_ptr;
-typedef std::unique_ptr<EC_POINT, void (*)(EC_POINT*)> EC_POINT_ptr;
-typedef std::unique_ptr<EC_KEY, void (*)(EC_KEY*)> EC_KEY_ptr;
-typedef std::unique_ptr<ECDSA_SIG, void (*)(ECDSA_SIG*)> ECDSA_SIG_ptr;
-// Error handling
 namespace Error = pdo::error;
 
 // XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
-// Utility function: Deserialize ECDSA Private Key
-// throws RuntimeError, ValueError
-EC_KEY* deserializeECDSAPrivateKey(const std::string& encoded)
+// Default constructor, custom curve constructor
+// PDO_DEFAULT_SIGCURVE is define that must be provided at compile time
+// Its default value is set in the cmake file
+pcrypto::sig::PrivateKey::PrivateKey(const pcrypto::sig::SigCurve& sigCurve)
 {
-    BIO_ptr bio(BIO_new_mem_buf(encoded.c_str(), -1), BIO_free_all);
-    if (!bio)
-    {
-        std::string msg("Crypto Error (deserializeECDSAPrivateKey): Could not create BIO");
-        throw Error::RuntimeError(msg);
-    }
-
-    EC_KEY* private_key = PEM_read_bio_ECPrivateKey(bio.get(), NULL, NULL, NULL);
-    if (!private_key)
-    {
-        std::string msg(
-            "Crypto Error (deserializeECDSAPrivateKey): Could not "
-            "deserialize private ECDSA key");
-        throw Error::ValueError(msg);
-    }
-    return private_key;
-}  // deserializeECDSAPrivateKey
+    key_ = nullptr;
+    if (sigCurve == pcrypto::sig::SigCurve::UNDEFINED)
+        sigDetails_ = pcrypto::sig::SigDetails[static_cast<int>(SigCurve::PDO_DEFAULT_SIGCURVE)];
+    else
+        sigDetails_ = pcrypto::sig::SigDetails[static_cast<int>(sigCurve)];
+}  // pcrypto::sig::PrivateKey::PrivateKey
 
 // XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
-// Default constructor (default curve specified in PDO_DEFAULT_SIGCURVE)
-pcrypto::sig::PrivateKey::PrivateKey()
+// Custom curve constructor with initial key specified as a bignum
+pcrypto::sig::PrivateKey::PrivateKey(
+    const pcrypto::sig::SigCurve& sigCurve,
+    const ByteArray& numeric_key) :
```

Comment:
i could go either way... its numeric to me because the code i'm writing around it is treating it as a number (hierarchical key derivation). public keys make more sense because they are compressed points.

Reply:
well, even the string/PEM version is still an (encoded) number?   I definitely wouldn't want to veto this name but it just wasn't very intuitive for me ...